### ISSUE TITLE
InstancedNode : get exceptions on Android with jme3.1-stable

### ISSUE THREAD

**Issue by jmecn** [jMonkeyEngine/jmonkeyengine#635]
Just tried jme3test.scene.instance.TestInstanceNode on Android. I get two exceptions.

** Exception 1**

    com.jme3.app.AndroidHarness: SEVERE Exception thrown in Thread[GLThread 183,5,main]
    com.jme3.renderer.RendererException: compile error in: ShaderSource[name=Common/MatDefs/Misc/Unshaded.vert, defines, type=Vertex, language=GLSL100]
                               ERROR: 0:158: '=' : global variable initializers must be constant expressions 

                                   at com.jme3.renderer.opengl.GLRenderer.updateShaderSourceData(GLRenderer.java:1200)
                                   at com.jme3.renderer.opengl.GLRenderer.updateShaderData(GLRenderer.java:1227)
                                   at com.jme3.renderer.opengl.GLRenderer.setShader(GLRenderer.java:1291)
                                   at com.jme3.material.logic.DefaultTechniqueDefLogic.render(DefaultTechniqueDefLogic.java:94)
                                   at com.jme3.material.Technique.render(Technique.java:166)
                                   at com.jme3.material.Material.render(Material.java:968)
                                   at com.jme3.renderer.RenderManager.renderGeometry(RenderManager.java:616)
                                   at com.jme3.renderer.queue.RenderQueue.renderGeometryList(RenderQueue.java:266)
                                   at com.jme3.renderer.queue.RenderQueue.renderQueue(RenderQueue.java:305)
                                   at com.jme3.renderer.RenderManager.renderViewPortQueues(RenderManager.java:870)
                                   at com.jme3.renderer.RenderManager.flushQueue(RenderManager.java:781)
                                   at com.jme3.renderer.RenderManager.renderViewPort(RenderManager.java:1097)
                                   at com.jme3.renderer.RenderManager.render(RenderManager.java:1145)
                                   at com.jme3.app.SimpleApplication.update(SimpleApplication.java:253)
                                   at com.jme3.app.AndroidHarness.update(AndroidHarness.java:497)
                                   at com.jme3.system.android.OGLESContext.onDrawFrame(OGLESContext.java:336)
                                   at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1548)
                                   at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1259)

I checked the source, maybe something wrong in `Common/ShadeLib/Instance.glsllib`, line 43:

    mat4 worldMatrix = mat4(vec4(inInstanceData[0].xyz, 0.0), 
                            vec4(inInstanceData[1].xyz, 0.0), 
                            vec4(inInstanceData[2].xyz, 0.0), 
                            vec4(inInstanceData[3].xyz, 1.0));

because it said that `ERROR: 0:158: '=' : global variable initializers must be constant expressions `

** Exception 2**

    com.jme3.util.BufferUtils: SEVERE {0}
                             java.lang.IllegalArgumentException: Expected receiver of type sun.nio.ch.DirectBuffer, but got java.nio.ByteBufferAsFloatBuffer
                                 at java.lang.reflect.Method.invoke(Native Method)
                                 at com.jme3.util.BufferUtils.destroyDirectBuffer(BufferUtils.java:1330)
                                 at com.jme3.scene.instancing.InstancedGeometry.setMaxNumInstances(InstancedGeometry.java:196)
                                 at com.jme3.scene.instancing.InstancedGeometry.addInstance(InstancedGeometry.java:322)
                                 at com.jme3.scene.instancing.InstancedNode.addToInstancedGeometry(InstancedNode.java:239)
                                 at com.jme3.scene.instancing.InstancedNode.instance(InstancedNode.java:293)
                                 at com.jme3.scene.instancing.InstancedNode.instance(InstancedNode.java:300)
                                 at com.jme3.scene.instancing.InstancedNode.instance(InstancedNode.java:306)
                                 at net.jmecn.mygame.TestInstanceNode.simpleInitApp(TestInstanceNode.java:94)
                                 at com.jme3.app.SimpleApplication.initialize(SimpleApplication.java:220)
                                 at com.jme3.app.AndroidHarness.initialize(AndroidHarness.java:478)
                                 at com.jme3.system.android.OGLESContext.onDrawFrame(OGLESContext.java:328)
                                 at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1548)
                                 at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1259)

I checked the source of com.jme3.util.BufferUtils.java, the bug seems has been fixed.

https://github.com/jMonkeyEngine/jmonkeyengine/commit/848a9217d04184ebda03544cdaed2190d1d4f2c7

**Comment by shadowislord** [comment]
Where is the shader source that's supposed to be printed out above the exception?
It's not possible to debug this error without the shader source.

**Comment by Ali-RS** [comment]
More info available on the forum: 
https://hub.jmonkeyengine.org/t/exception-when-destroying-direct-memory-in-android/42483

**Comment by Ali-RS** [comment]
Should be fixed once PR #1214 is merged.

**Comment by joliver82** [comment]
There could be an error in the first lines of MorphAnim.glsllib included in Unshaded.vert:

```
    vec3 dummy_norm = vec3(0.0);
    vec3 dummy_tan = vec3(0.0);
```

because GLSL doesn't allow to assign a value of a globar variable when defining it but there's no NUM_MORPH_TARGETS defined in this testcase. Nowadays jme3test.scene.instance.TestInstanceNode works after proper instancing support has been added to android in PR https://github.com/jMonkeyEngine/jmonkeyengine/pull/1147

** Side note, we should check morph in android also.

About the second exception, it's covered in PR https://github.com/jMonkeyEngine/jmonkeyengine/pull/1214

**Comment by Ali-RS** [comment]
I am closing this as the instancing issue is fixed in the master branch. 
Regarding MorphAnim.glsllib I think we need to open a new issue.
